layers/ipfix.go

```diff
@@ -0,0 +1,773 @@
+	// A value of 3 is reserved for Options Template Sets. Values from 4
+	// to 255 are reserved for future use. Values 256 and above are used
+	// for Data Sets. The Set ID values of 0 and 1 are not used, for
+	// historical reasons [RFC3954].
+	ID IPFIXSetIDType
+	// Total length of the Set, in octets, including the Set Header, all
+	// records, and the optional padding. Because an individual Set MAY
+	// contain multiple records, the Length value MUST be used to
+	// determine the position of the next Set.
+	Length uint16
+}
+
+// Len returns the length in octet requires by Set header
+func (sh IPFIXSetHeader) Len() int {
+	return 2 + 2
+}
+
+func (sh *IPFIXSetHeader) decodeFromBytes(data *[]byte) error {
+	if len(*data) < sh.Len() {
+		*data = []byte{}
+		return fmt.Errorf("invalid set header length")
+	}
+
+	*data, sh.ID = (*data)[2:], IPFIXSetIDType(binary.BigEndian.Uint16((*data)[:2]))
+	*data, sh.Length = (*data)[2:], binary.BigEndian.Uint16((*data)[:2])
+
+	switch {
+	case sh.ID <= 1:
+		return fmt.Errorf("invalid set header, ID %d reserved for legacy", sh.ID)
+	case sh.ID >= 4 && sh.ID <= 255:
+		return fmt.Errorf("invalid set header, ID %d reserved for future use", sh.ID)
+	}
+
+	if len(*data) < int(sh.Length)-sh.Len() {
+		return fmt.Errorf("invalid data length")
+	}
+
+	return nil
+}
+
+// IPFIXTemplateRecord contains any combination of IANA-assigned and/or
+// enterprise-specific Information Element identifiers (RFC 7011 section 3.4.1)
+//
+//    0                   1                   2                   3
+//    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |      Template ID (> 255)      |         Field Count           |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   | Field Specifiers                                              |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   ...
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   | Field Specifiers                                              |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//
+//   The struct also contains the set header for convenience.
+type IPFIXTemplateRecord struct {
+	// Set header
+	Header IPFIXSetHeader
+	// Each Template Record is given a unique Template ID in the range
+	// 256 to 65535.  This uniqueness is local to the Transport Session
+	// and Observation Domain that generated the Template ID.
+	ID IPFIXSetIDType
+	// Number of Field Specifiers in this Template Record.
+	FieldCount uint16
+	// List of Field Specifiers in that template record.
+	// Order matter here, as content will be decoded with that order.
+	Fields []IPFIXFieldSpecifier
+
+	// for each template compute the minimum length an associated data record
+	// can contains. That helps to iter over multiple data records in the same
+	// set.
+	minDataRecordLength uint16
```

Comment:
minTemplateRecordLength

Reply:
that is the minimum length of a data record which follow that template, not the length of the template record